rhiever/tpot
# Operator.inheritors() returns a list rather than a set.

## What does this PR do?
In the current version of TPOT, different pipelines are generated, even if the same random state is given. This PR aims to remedy that.

When picking a new primitive when generating a new individual, this is done by calling `np.random.choice(pset.terminals[type_])`. While numpy properly had its seed set, `pset.terminals` was not ordered - resulting in a different primitive picked anyway.

The reason `pset.terminals` was unordered, is that it is constructed in the same order as  `Operator.inheritors()` returns the available operators. However, this function internally stored the operators in a `set`, which is unordered in Python. Hence, multiple calls of `Operator.inheritors()` could result in differently ordered sets, meaning the random choice, while always picking the same index, would pick different operators.

This problem is fixed by storing the operators in a `list`, since that remembers the order elements were added to it. 

Edit: Apparently on my windows system, it was not enough to just use a list, but the list had to be explicitly sorted as well, this change has been made in this PR. On my Linux (Ubuntu) systems it worked fine without.

## Where should the reviewer start?
All actual code changes are [here](https://github.com/PG-TUe/tpot/blob/ordered-operators/tpot/operators/base.py#L125#L135).

The new unit test is [here](https://github.com/PG-TUe/tpot/commit/59e18aedfba4331e993f64a3c85b69bd9a92b8fc).

## How should this PR be tested?
To test if `Operator.inheritors()` now always returns the same list, simply call it a few times and see if it indeed works now. This is also done in the unit test. Unfortunately I don't know of a way to better test this change (note that the old code *could* also return the same order).

## Any background context you want to provide?
I feel enough background was given.

## What are the relevant issues?
#349

## Questions:
I feel that the docs do not need to be updated, as to the best of my knowledge it is intended that the results should always be reproducible, this is a bugfix.

It adds no new dependencies.
